### PR TITLE
bpo-46031: add POP_JUMP_IF_NOT_NONE and POP_JUMP_IF_NONE

### PR DESCRIPTION
**Pull request by penguin-wwy**
https://bugs.python.org/issue46031

https://github.com/faster-cpython/ideas/discussions/154

### CODE REVIEW

File: Python/ceval.c [hunk 1]
```diff
@@ -4052,6 +4052,30 @@ _PyEval_EvalFrameDefault(PyThreadState *tstate, InterpreterFrame *frame, int thr
             DISPATCH();
         }
 
+        TARGET(POP_JUMP_IF_NOT_NONE) {
+            PyObject *value = POP();
+            if (value != Py_None) {
```
Comment (brandtbucher) [inline Python/ceval.c]:
```suggestion
            if (!Py_IsNone(value)) {
```

File: Python/ceval.c [hunk 2]
```diff
@@ -4052,6 +4052,30 @@ _PyEval_EvalFrameDefault(PyThreadState *tstate, InterpreterFrame *frame, int thr
...
+                CHECK_EVAL_BREAKER();
+                DISPATCH();
+            }
+            Py_DECREF(value);
+            DISPATCH();
+        }
+
+        TARGET(POP_JUMP_IF_NONE) {
+            PyObject *value = POP();
+            if (value == Py_None) {
```
Comment (brandtbucher) [inline Python/ceval.c]:
```suggestion
            if (Py_IsNone(value)) {
```

File: Lib/opcode.py [hunk 3]
```diff
@@ -150,6 +150,9 @@ def jabs_op(name, op):
 def_op('DELETE_FAST', 126)      # Local variable number
 haslocal.append(126)
 
+jabs_op('POP_JUMP_IF_NOT_NONE', 127)
+jabs_op('POP_JUMP_IF_NONE', 128)
+
```
Comment (brandtbucher) [inline Lib/opcode.py]:
```suggestion
jabs_op('POP_JUMP_IF_NOT_NONE', 127)
jabs_op('POP_JUMP_IF_NONE', 128)
```

File: Misc/NEWS.d/next/Core and Builtins/2021-12-10-09-10-32.bpo-46031.rM7JOX.rst [hunk 4]
```diff
@@ -0,0 +1 @@
+Add ``POP_JUMP_IF_NOT_NONE`` and ``POP_JUMP_IF_NONE`` opcode to speed up conditional jump.
```
Comment (brandtbucher) [inline Misc/NEWS.d/next/Core and Builtins/2021-12-10-09-10-32.bpo-46031.rM7JOX.rst]:
```suggestion
Add :opcode:`POP_JUMP_IF_NOT_NONE` and :opcode:`POP_JUMP_IF_NONE` opcodes to speed up conditional jumps.
```

Can you also update [What's New](https://github.com/python/cpython/blob/main/Doc/whatsnew/3.11.rst#cpython-bytecode-changes), too?

File: Python/compile.c [hunk 5]
```diff
@@ -8159,6 +8161,18 @@ optimize_basic_block(struct compiler *c, basicblock *bb, PyObject *consts)
                             bb->b_instr[i+1].i_opcode = NOP;
                         }
                         break;
+                    case IS_OP:
+                        cnt = get_const_value(inst->i_opcode, oparg, consts);
```
Comment (brandtbucher) [inline Python/compile.c:8525]:
```suggestion
                        cnt = get_const_value(inst->i_opcode, oparg, consts);
                        if (cnt == NULL) {
                            goto error;
                        }
```

File: Python/compile.c [hunk 6]
```diff
@@ -8159,6 +8161,18 @@ optimize_basic_block(struct compiler *c, basicblock *bb, PyObject *consts)
...
+                        int jump_op = i+2 < bb->b_iused ? bb->b_instr[i+2].i_opcode : 0;
+                        if (Py_IsNone(cnt) && (jump_op == POP_JUMP_IF_FALSE || jump_op == POP_JUMP_IF_TRUE)) {
+                            Py_DECREF(cnt);
+                            unsigned char nextarg = bb->b_instr[i+1].i_oparg;
+                            inst->i_opcode = NOP;
+                            bb->b_instr[i+1].i_opcode = NOP;
+                            bb->b_instr[i+2].i_opcode = nextarg ^ (jump_op == POP_JUMP_IF_FALSE) ?
+                                    POP_JUMP_IF_NOT_NONE : POP_JUMP_IF_NONE;
+                        }
```
Comment (brandtbucher) [inline Python/compile.c]: Pretty sure you leak `cnt` when it's non-`None`:

```suggestion
                            unsigned char nextarg = bb->b_instr[i+1].i_oparg;
                            inst->i_opcode = NOP;
                            bb->b_instr[i+1].i_opcode = NOP;
                            bb->b_instr[i+2].i_opcode = nextarg ^ (jump_op == POP_JUMP_IF_FALSE) ?
                                    POP_JUMP_IF_NOT_NONE : POP_JUMP_IF_NONE;
                        }
                        Py_DECREF(cnt);
```

**Review by brandtbucher — CHANGES_REQUESTED** [review]
Thanks for the patch @penguin-wwy, and welcome to CPython!

File: Python/compile.c (hunk 6 above)
Comment (brandtbucher) [reply Python/compile.c]: That's an annoying diff. All I did was move the `Py_DECREF`.

Comment (penguin-wwy) [reply Python/compile.c]: Thanks for the reminder, I overlooked `Py_INCREF` in `get_const_value`

File: Misc/NEWS.d/next/Core and Builtins/2021-12-10-09-10-32.bpo-46031.rM7JOX.rst (hunk 4 above)
Comment (penguin-wwy) [reply Misc/NEWS.d/next/Core and Builtins/2021-12-10-09-10-32.bpo-46031.rM7JOX.rst]: Update now

**Comment by penguin-wwy** [comment]
I have made the requested changes; please review again.

**Review by brandtbucher — APPROVED** [review]
@markshannon, can you hit this with `pyperformance` to see what the numbers look like? Hopefully we can reproduce the reported [1% speedup](https://github.com/faster-cpython/ideas/discussions/154#discussioncomment-1784345).

**Review by brandtbucher — CHANGES_REQUESTED** [review]
Ah, I didn't see you have some stale files. @penguin-wwy, can you run `make regen-all`, please?

**Comment by penguin-wwy** [comment]
@brandtbucher I resolve conflicts and run `make regen-all`,  then nothing reminds me that I need to submit news.
@bedevere-bot I have made the requested changes; please review again.

**Comment by markshannon** [comment]
Shows a [small speedup](https://gist.github.com/markshannon/a1d71a5d61d344da20bb063302dfa5c6)

**Comment by penguin-wwy** [comment]
@markshannon Thank you. Also, what are the configuration arguments for exec benchmark ? I used `--enable-optimizations --with-lto`

**Comment by markshannon** [comment]
> @markshannon Thank you. Also, what are the configuration arguments for exec benchmark ? I used `--enable-optimizations --with-lto`

Yes, `--enable-optimizations --with-lto` is what we use.
Different builds and architectures can impact the results quite a lot, so it is quite likely that one machine will show a speedup of 1% and another, little or no speedup.

**Comment by markshannon** [comment]
Looks good.
You will need to add the new instructions to `Doc/library/dis.rst`
and increment the magic number in `Lib/importlib/_bootstrap_external.py`

**Comment by penguin-wwy** [comment]
Thank you! I have made the requested changes.

File: Lib/importlib/_bootstrap_external.py [hunk 7]
```diff
@@ -374,6 +374,7 @@ def _write_atomic(path, data, mode=0o666):
 #     Python 3.11a3 3466 (bpo-45292: PEP-654 except*)
 #     Python 3.11a4 3467 (Change CALL_xxx opcodes)
 #     Python 3.11a4 3468 (Add SEND opcode)
+#     Python 3.11a4 3469 (Add POP_JUMP_IF_NOT_NONE/POP_JUMP_IF_NONE opcode)
```
Comment (markshannon) [inline Lib/importlib/_bootstrap_external.py]: `opcode` -> `opcodes`

**Comment by markshannon** [comment]
LGTM.
I'm afraid that there are a few PRs adding or removing bytecodes at the moment, so you might need to wait a few days to avoid conflicts.

**Comment by penguin-wwy** [comment]
You are right 
I have solved conflicts >= 5 times :)

**Pull request merged**